src/Buffer.mo

```diff
@@ -1,166 +1,1764 @@
+  public func isSubBufferOf<X>(subBuffer : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    switch (indexOfBuffer(subBuffer, buffer, equal)) {
+      case null subBuffer.size() == 0;
+      case _ true;
+    };
+  };
+
+  /// Checks if `subBuffer` is a strict subBuffer of `buffer`, i.e. `subBuffer` must be
+  /// strictly contained inside both the first and last indices of `buffer`.
+  /// Uses `equal` to compare elements.
+  ///
+  /// Runtime: O(size of subBuffer + size of buffer)
+  ///
+  /// Space: O(size of subBuffer)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func isStrictSubBufferOf<X>(subBuffer : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    let subBufferSize = subBuffer.size();
+
+    switch (indexOfBuffer(subBuffer, buffer, equal)) {
+      case (?index) {
+        index != 0 and index != (buffer.size() - subBufferSize : Nat) // enforce strictness
+      };
+      case null {
+        subBufferSize == 0 and subBufferSize != buffer.size();
+      };
+    };
+  };
+
+  /// Returns the prefix of `buffer` of length `length`. Traps if `length`
+  /// is greater than the size of `buffer`.
+  ///
+  /// Runtime: O(length)
+  ///
+  /// Space: O(length)
+  public func prefix<X>(buffer : Buffer<X>, length : Nat) : Buffer<X> {
+    let size = buffer.size();
+    if (length > size) {
+      Prim.trap "Buffer index out of bounds in prefix";
+    };
+
+    let newBuffer = Buffer<X>(newCapacity length);
+
+    var i = 0;
+    while (i < length) {
+      newBuffer.add(buffer.get(i));
+      i += 1;
+    };
+
+    newBuffer;
+  };
+
+  /// Checks if `prefix` is a prefix of `buffer`. Uses `equal` to
+  /// compare elements.
+  ///
+  /// Runtime: O(size of prefix)
+  ///
+  /// Space: O(size of prefix)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func isPrefixOf<X>(prefix : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    let sizePrefix = prefix.size();
+    if (buffer.size() < sizePrefix) {
+      return false;
+    };
+
+    var i = 0;
+    while (i < sizePrefix) {
+      if (not equal(buffer.get(i), prefix.get(i))) {
+        return false;
+      };
+
+      i += 1;
+    };
+
+    return true;
+  };
+
+  /// Checks if `prefix` is a strict prefix of `buffer`. Uses `equal` to
+  /// compare elements.
+  ///
+  /// Runtime: O(size of prefix)
+  ///
+  /// Space: O(size of prefix)
+  ///
+  /// *Runtime and space assumes that `equal` runs in O(1) time and space.
+  public func isStrictPrefixOf<X>(prefix : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
+    let sizePrefix = prefix.size();
+    if (buffer.size() <= sizePrefix) {
+      return false;
+    };
+
+    var i = 0;
+    while (i < sizePrefix) {
+      if (not equal(buffer.get(i), prefix.get(i))) {
+        return false;
+      };
+
+      i += 1;
+    };
+
+    return true;
+  };
```

Comment:
Maybe, code simplification: 

```
  public func isStrictPrefixOf<X>(prefix : Buffer<X>, buffer : Buffer<X>, equal : (X, X) -> Bool) : Bool {
    if (buffer.size() <= prefix.size()) {
      return false;
    };
    isPrefixOf(prefix, buffer, equal);
  };
  ```

Reply:
Ah right. I was initially trying to remove the overhead of the function call but in hind sight that seems like over-optimizing. Will refactor as you're suggesting.

Reply:
Mm I guess the downside is what @ggreif was mentioning in the walk-through, that `equal` will get further from its definition, and so the compiler may have a harder time optimizing it if `equal` is a constant function.

Reply:
I would go for the simpler one.